### ISSUE TITLE
YOLO Features to ONNX

### ISSUE THREAD

**Issue by Djkmarco** [ultralytics/ultralytics#4062]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello! I'm looking for a suggestion:

I've a pretrained YOLOv8n model. I want to add hooks to extract features, doing few calculations, and output YOLO results along with processed features. All of this should then be exported to ONNX.

Just to give you a snapshot:

```
model = YOLO("pretrained.pt")
model = model.model
x = torch.rand(1, 3, 512, 1024)

class modified_yolo(nn.Module):
    def __init__(self, model):
        super(modified_yolo, self).__init__()
        self.use_cuda = torch.cuda.is_available()
        self.device = torch.device("cuda" if self.use_cuda else "cpu")
        if self.use_cuda:
            torch.cuda.manual_seed_all(1024)
        self.model = model
        self.model.to(self.device)

    random.seed(1024)
    torch.manual_seed(1024)

    def forward(self, x):
        output = []
        test_outputs = OrderedDict([("layer1", []), ("layer2", []), ("layer3", [])])

        model = self.model
        model.to(self.device)

        def hook(module, input, output):
            outputs.append(output)

        model.model[4].register_forward_hook(hook)
        model.model[6].register_forward_hook(hook)
        model.model[8].register_forward_hook(hook)

        # get results and features
        with torch.no_grad():
            yolo_result = model(x.to(self.device))
        for k, v in zip(test_outputs.keys(), outputs):
            test_outputs[k].append(v.cpu().detach())
        outputs = []
        for k, v in test_outputs.items():
            test_outputs[k] = torch.cat(v, 0)
        embedding_vectors = test_outputs["layer1"]
        for layer_name in ["layer2", "layer3"]:
            embedding_vectors = embedding_concat(
                embedding_vectors, test_outputs[layer_name]
            )

        return yolo_result, embedding_vectors

net = modified_yolo(model=model)

yolo_result, embedding_vectors = net(x)

```
The code do what it is supposed to do, but when exporting to onnx it get stuck.

```

torch.onnx.export(
    net,
    x,
    "modified_yolo.onnx",
    verbose=True,
    input_names=["image"],
    output_names=["yolo_result", "score_map"],
)
```

I have the same problem trying to export just the model using:

```

model = YOLO("pretrained.pt")
torch.onnx.export(model.model, x, "yolo.onnx", verbose=True, input_names=["image"])


```
Do you have any suggestion not necessarily using the same approach as mine?

**Comment by glenn-jocher** [comment]
@Djkmarco hello,

Firstly, thank you for providing such a detailed description of your issue. It does appear that you're aiming to not only perform object detection but also extract personalized feature maps from different layers, which is a task that requires careful handling when exporting to ONNX.

Your current situation could possibly be due to several factors such as unsupported operations in ONNX, dynamic shapes, or issues related to the hook functionality. 

When exporting the model to the ONNX format, please ensure that all the operations you're using in your modified network are supported by ONNX. Some PyTorch operators don't have a straightforward mapping in ONNX, and may be the cause of your export issue. 

Another important thing to note is that ONNX prefers defined tensor shapes. Try to make sure the tensor shapes in your model do not have dynamic sizes when you are performing the export. 

Lastly, ONNX might have a problem processing the hook functions, especially when they are trying to modify the computation graph. Registering hooks to extract intermediate layer outputs and include them in the model output might cause problems with the ONNX export. 

Without diving into specific code, it's hard to provide more specific solutions. However, I suggest you to divide and conquer: try to simplify your model as much as possible, get that exporting, and then gradually add complexity while ensuring that it can still export to ONNX at each step. 

I hope you find these pointers helpful and hopefully you can find a suitable solution that fulfills your requirements!

**Comment by Djkmarco** [comment]
Thank you @glenn-jocher!

Just one more thing... Is there a reason why 

`
model = YOLO("pretrained.pt")
torch.onnx.export(model.model, x, "yolo.onnx", verbose=True, input_names=["image"])
`

(pretrained model is a not modified yolov8 model) should not work?

**Comment by glenn-jocher** [comment]
Hello again, @Djkmarco!

The code you've posted for exporting a pretrained YOLOv8 model to ONNX seems to be fine at first glance. But, it's important to note that there might be some operations in the model that ONNX doesn't yet support. As you know, not every operation in PyTorch has a direct equivalent in ONNX, and this could lead to issues during the conversion process.

Also, please ensure that you've specified the correct input size `x` corresponding to the model you trained. This is necessary as ONNX requires fixed tensor sizes and doesn't handle dynamic input shapes well.

Also, do check if all the dependencies, Torch and ONNX versions, are up-to-date. Sometimes, mismatches in versions or outdated packages might cause some issues.

Without more detailed error information, it's a bit tricky to provide a concrete solution. But these are the most common reasons why the ONNX export can fail. 

In case of an error, please provide the error message for more productive troubleshooting. 

Good luck with your work!

**Comment by Djkmarco** [comment]
Thank you @glenn-jocher.

I'm using
ultralytics==8.0.124
onnx==1.14.0
torch==2.0.1

Simple reproducing code:
```

from ultralytics import YOLO
model = YOLO('yolov8s.pt')
model.export(format='onnx')

# this works in 0.8 seconds

import torch
x = torch.rand(1, 3, 640, 640)
torch.onnx.export(model.model, x, "yolo.onnx", verbose=True, input_names=["image"])

# this keeps running for an infinite time
```

**Comment by glenn-jocher** [comment]
Hi @Djkmarco,

It appears that you're experiencing a problem when trying to export your model to ONNX using the PyTorch's `torch.onnx.export()` function. The reason this call runs indefinitely may stem from the complexity of the object detection model and how it's been built.

YOLO v8, like many advanced object detection models, has a complex architecture with many layers and operations. When you're trying to export it with `torch.onnx.export()`, PyTorch needs to trace through all the operations in the model, which can sometimes lead to unexpected issues. 

Without a specific error message, it's hard to say definitely, but it could be that the export is getting caught in a loop or is struggling with some of the operations in the model. Remember that not all PyTorch operations are supported by ONNX, so you may encounter issues when using this function directly.

Instead, consider using the `export()` functionality provided in the Ultralytics library. This function has been crafted specifically to handle all the complexities of the Ultralytics models, taking care of all the specifics related to layer operations and architecture intricacies.

Please do update your versions of Ultralytics, ONNX, and PyTorch to the latest ones if not already done, as they have better support and fixes for many known issues.

So, in short, instead of using PyTorch's export, please use the export function provided by Ultralytics for ONNX. This should handle all the intricacies of the model and complete in a timely manner.

Happy coding!
